code/datums/components/trader/trader.dm

```diff
@@ -0,0 +1,413 @@
+/datum/component/trader/proc/on_attack_hand(atom/source, mob/living/carbon/customer)
+	SIGNAL_HANDLER
+	var/mob/living/living_trader = parent
+	if(!can_trade(customer) || customer.combat_mode)
+		return
+	var/list/npc_options = list()
+	if(products.len)
+		npc_options["Buy"] = radial_icons_cache[TRADER_RADIAL_BUY]
+	if(wanted_items.len)
+		npc_options["Sell"] = radial_icons_cache[TRADER_RADIAL_SELL]
+	if(length(trader_data.say_phrases))
+		npc_options["Talk"] = radial_icons_cache[TRADER_RADIAL_TALK]
+	if(!npc_options.len)
+		return
+
+	living_trader.face_atom(customer)
+
+	INVOKE_ASYNC(src, PROC_REF(open_npc_options), customer, npc_options)
+
+	return COMPONENT_CANCEL_ATTACK_CHAIN
+
+/**
+ * Generates a radial of the initial radials of the NPC
+ * Called via asynch, due to the sleep caused by show_radial_menu
+ * Arguments:
+ * * customer - (Mob REF) The mob trying to buy something
+ */
+/datum/component/trader/proc/open_npc_options(mob/living/carbon/customer, list/npc_options)
+	if(!can_trade(customer))
+		return
+	var/npc_result = show_radial_menu(customer, parent, npc_options, custom_check = CALLBACK(src, PROC_REF(check_menu), customer), require_near = TRUE, tooltips = TRUE)
+	switch(npc_result)
+		if("Buy")
+			buy_item(customer)
+		if("Sell")
+			try_sell(customer)
+		if("Talk")
+			discuss(customer)
+
+/**
+ * Checks if the customer is ok to use the radial
+ *
+ * Checks if the customer is not a mob or is incapacitated or not adjacent to the source of the radial, in those cases returns FALSE, otherwise returns TRUE
+ * Arguments:
+ * * customer - (Mob REF) The mob checking the menu
+ */
+/datum/component/trader/proc/check_menu(mob/customer)
+	if(!istype(customer))
+		return FALSE
+	if(IS_DEAD_OR_INCAP(customer) || !customer.Adjacent(parent))
+		return FALSE
+	return TRUE
+
+/**
+ * Generates a radial of the items the NPC sells and lets the user try to buy one
+ * Arguments:
+ * * customer - (Mob REF) The mob trying to buy something
+ */
+/datum/component/trader/proc/buy_item(mob/customer)
+	if(!can_trade(customer))
+		return
+	if(!LAZYLEN(products))
+		return
+	var/mob/living/trader = parent
+	var/list/display_names = list()
+	var/list/items = list()
+	var/list/product_info
+	for(var/obj/item/thing as anything in products)
+		display_names["[initial(thing.name)]"] = thing
+
+		if(!radial_icons_cache[thing])
+			radial_icons_cache[thing] = image(icon = initial(thing.icon), icon_state = initial(thing.icon_state_preview) ? initial(thing.icon_state_preview) : initial(thing.icon_state))
+
+		var/image/item_image = radial_icons_cache[thing]
+		product_info = products[thing]
+		if(product_info[TRADER_PRODUCT_INFO_QUANTITY] <= 0) //out of stock
+			item_image.overlays += radial_icons_cache[TRADER_RADIAL_OUT_OF_STOCK]
+		items += list("[initial(thing.name)]" = item_image)
+	var/pick = show_radial_menu(customer, trader, items, custom_check = CALLBACK(src, PROC_REF(check_menu), customer), require_near = TRUE, tooltips = TRUE)
```

Comment:
Similar thing here for "what if they only have one product"?
Maybe I should ask _you_ to make `show_radial_menu` just automatically return a result instead of displaying if it is provided with a list of one option 😅